distributed/utils.py

```diff
@@ -76,3 +81,80 @@ class NoColor:
     cyan = ""
     white = ""
     reset = ""
+
+
+class TrackTime:
+    """integrated class for perf timing via perf_counter"""
+
+    def __init__(self, use_ms: bool = False, round_to: Optional[int] = 4):
+        self.use_ms = use_ms
+        self.round_to = round_to
+        self.start_time = 0.0
+        self.elapsed_time = 0.0
+        self.unit = "seconds" if not use_ms else "milliseconds"
+
+    def __enter__(self):
+        self.start_time = time.perf_counter()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        end_time = time.perf_counter()
+        self.elapsed_time = end_time - self.start_time
+
+        if self.use_ms:
+            self.elapsed_time *= 1000  # Convert to milliseconds
+
+        if self.round_to is not None:
+            self.elapsed_time = round(self.elapsed_time, self.round_to)
+
+    def get_time(self) -> float:
+        return self.elapsed_time
+
+
+class CUDATrackTime:
+    """integrated class for perf timing via cuda events"""
+
+    def __init__(self, use_ms: bool = False, round_to: Optional[int] = 4):
+        self.start_event = torch.cuda.Event(enable_timing=True)
+        self.end_event = torch.cuda.Event(enable_timing=True)
+        self.active = False
+        self.round_to = round_to
+        self.elapsed_time = 0.0
+        self.use_ms = use_ms
+        self.unit = "seconds" if not use_ms else "milliseconds"
+
+    def start(self):
+        if self.active:
+            raise RuntimeError("Timer is already running. Use .stop() to stop it")
+        self.start_event.record()
+        self.active = True
+
+    def stop(self):
+        if not self.active:
+            raise RuntimeError("Timer is not running. Use .start() to start it")
+        self.end_event.record()
+        self.active = False
```

Comment:
nit: can you add a note here that these methods (containing `.record()`) can create CUDA context on `current_device()` (whatever it is) if there hasn't been one? Thus for safety, `torch.cuda.set_device()` is strongly needed for `CUDATrackTime`.

Reply:
I've updated this to take a device and then to use that for setting the events.  
Also added comment re: creating context on the designated device.